keps/sig-cli/2206-openapi-features-in-kustomize.md

```diff
@@ -0,0 +1,492 @@
+- [ ] Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+OpenAPI is an open specification for defining types and operations on those 
+types.  The Kubernetes API server has supported it fully since version 1.15,
+meaning one can query an API server to get a detailed list of the types it 
+understands, including any Custom Resources that it knows about.  Clients can 
+use this information to correctly instantiate those types and make calls on 
+the server.
+
+Kustomize uses a hard-coded snapshot of the kubernetes API taken at version 
+1.19 to help it understand field types and most importantly, merge keys to 
+use when patching type instances.  This is how, for example, kustomize can 
+change the `image` name used in a particular container in a particular 
+Deployment instance.
+
+Two immediate problems with this setup are 1), to upgrade the OpenAPI version, 
+once must recompile kustomize with a new snapshot, 2) there’s no simple way to 
+teach kustomize about a custom resource (field types, field validation rules, 
+merge keys), despite the availability of OpenAPI schema data from custom 
+resource tooling like kubebuilder.
+
+If a kustomize user wants to specify their own OpenAPI schema files to use. 
+Because the Kubernetes OpenAPI schema is currently built into the kustomize 
+binary, that strategic merging for custom resources is not supported well. 
+If a kustomize user were able to specify their own custom OpenAPI schema via 
+an `openapi` field in a kustomization file, they would be able to specify 
+patch strategy and merge key information for their custom resources. This 
+proposal describes what that openapi field would look like and how the user 
+would be able to get an OpenAPI schema containing their custom resource 
+information. 
+
+## Motivation
+
+Allowing users to specify their own OpenAPI schema files will give them more 
+flexibility in how kustomize does strategic merging, especially for custom 
+resources. Users will be able to choose the patch strategy and merge keys for 
+their resources. 
+
+For a short presentation, see https://docs.google.com/presentation/d/1p3ASgdGvLYoUVLmx60_rjLuwuz4ypzwgj2ZWp7-kCjQ/preview?slide=id.gb19ebdc14e_0_0
+
+### Goals
+
+- In the `openapi` field of a kustomization file, a user may specify a path 
```

Comment:
Consider making it a list for multiple schemas

Reply:
I think that's a good idea and we have discussed doing so in the future but I am planning to leave it as a scalar for now for simplicity